Work towards implementing browsing.scripting() web extension APIs.

kiaraarose opened the pull request:
#### 9cde205004ecef8ae7813407e7417b8d55135a69
<pre>
Work towards implementing browsing.scripting() web extension APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259954">https://bugs.webkit.org/show_bug.cgi?id=259954</a>

Reviewed by Timothy Hatcher.

This patch adds support for script injections with the scripting.insertCSS(), and scripting.removeCSS() APIs.

* Source/WTF/wtf/Ref.h:
(WTF::operator==):
Add support for operator== so that removeFirst(const U&amp;) works for Ref types.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
(WebKit::boolForKey):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h:
(WebKit::toWebExtensionFrameIdentifier):
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionResultParameters.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm: Added.
(WebKit::WebExtensionContext::scriptingExecuteScript):
(WebKit::WebExtensionContext::scriptingInsertCSS):
(WebKit::WebExtensionContext::scriptingRemoveCSS):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didCommitLoadForFrame):
Clear dynamicallyInjectedUserStyleSheets when a page navigation occurs since the stylesheets should
be removed when the user leaves or refreshes the page.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm: Added.
(WebKit::WebExtensionDynamicScripts::userStyleSheetMatchesContent):
(WebKit::WebExtensionDynamicScripts::populateFrames):
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
(WebKit::WebExtensionDynamicScripts::populateStyleSheetPairs):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::dynamicallyInjectedUserStyleSheets):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseTargetInjectionOptions):
(WebKit::parseScriptInjectionOptions):
(WebKit::parseCSSInjectionOptions):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::validateTarget):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/269018@main">https://commits.webkit.org/269018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e25cee73cb5ad9776a0a5e03a394ee6206308119

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21352 "Failed to checkout and rebase branch from PR 18507") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/22405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/23215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21592 "Failed to checkout and rebase branch from PR 18507") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/24965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/21913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21576 "Failed to checkout and rebase branch from PR 18507") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/24965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/22405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24067 "Built successfully") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/21535 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/24965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/22405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/18572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/24965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/22405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/23527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/20736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/21913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/24755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/22405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/5852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26025 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2643 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/19950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/5677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WTF/wtf/Ref.h:
```diff
@@ -208,6 +217,24 @@ inline Ref<T, U>& Ref<T, U>::operator=(const Ref<X, Y>& reference)
+    return a.m_ptr == b.m_ptr;
+}
+
+template<typename X, typename Y, typename W, typename Z>
```
`, typename Z` can be removed.

jyavenard replied:
that wasn't removed

xeenon commented on Source/WTF/wtf/Ref.h:
```diff
@@ -208,6 +217,24 @@ inline Ref<T, U>& Ref<T, U>::operator=(const Ref<X, Y>& reference)
+    return a.m_ptr == b;
+}
+
+template<typename X, typename Y, typename W, typename Z>
```
Ditto.

rreno commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    populateStyleSheetPairs(WTFMove(styleSheetPairs), m_extension, parameters.files, parameters.css);
+
+    UserStyleSheetVector styleSheetsToRemove;
+    for (auto& styleSheetContent : styleSheetPairs) {
```
You moved this variable up on line 165 so this is a valid, but moved from variable. It might not be doing what you think!

rreno replied:
"this variable" meaning `styleSheetPairs`

brentfulgham replied:
Yes -- I noticed that in the other code, too. We only want to use `WTFMove` to pass something to another method that we are finished with in the current context. Otherwise, we want to pass a copy (if we don't want local state affected), or a reference (if we want the calling context to see the changes to the argument when it returns).

rreno commented on Source/WTF/wtf/Ref.h:
```diff
@@ -139,6 +139,15 @@ class Ref {
     friend Ref adoptRef<T>(T&);
     template<typename X, typename Y> friend class Ref;
 
+    template<typename X, typename Y, typename W, typename Z>
```
nit: The pattern is T, U for lhs and X, Y for rhs elsewhere, would be nice to keep the consistency.

xeenon replied:
T conflicted when I write this for Kiara. See `typename X, typename Y` above.

rreno replied:
Right, so you can just do `template <typename X, typename Y> friend bool operator==(const Ref&, const Ref<X, Y>&);`

rreno replied:
to be clear the X Y above is so that every Ref instantiation is friended with every other Ref instantiation.

kiaraarose replied:
Was getting build errors for the first Ref not being templated. Left this as is, but open to suggestions on how to resolve.

rreno commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+    std::optional<Vector<String>> files;
+    std::optional<Vector<size_t>> frameIDs;
+
+    std::optional<String> css;
```
Can we encode nullopt as a null string here instead of using optionals? ditto line below.

xeenon replied:
I think it helps being explicit when css is used or not.

rreno commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionResultParameters.h:
```diff
@@ -0,0 +1,42 @@
+namespace WebKit {
+
+struct WebExtensionScriptInjectionResultParameters {
+    std::optional<String> error;
```
Same as before - can we use null strings to encode nullopt? std::optional<String> seems a bit redundant.

xeenon replied:
We are doing this to be explicit, but yeah String can be special.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -53,17 +57,91 @@
+    if (parameters.frameIDs)
+        parameters.allFrames = false;
+    else
+        parameters.allFrames = boolForKey(targetInfo, allFramesKey, true);
```
```
if (!parameters.frameIDs)
    parameters.allFrames = boolForKey(targetInfo, allFramesKey, true);
```

Assuming `parameters.allFrames` defaults to `false`. See other comment.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+    std::optional<String> func;
+
+    WebKit::WebExtensionContentWorldType world;
+    bool allFrames;
```
```
WebExtensionContentWorldType world { WebExtensionContentWorldType::ContentScript };
bool allFrames { false };
```

Note: drop the `WebKit::`,

xeenon replied:
Or drop `allFrames` altogether, see other comment about using `!frameIDs` to mean all.

brentfulgham commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -29,6 +29,7 @@
 #import <wtf/UUID.h>
 #import <wtf/WallTime.h>
 #import <wtf/text/StringHash.h>
+//#import <wtf/cocoa/VectorCocoa.h>
```
Whoops! Please delete before landing!

rreno replied:
Please remove dead code

xeenon replied:
Remove?

brentfulgham commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#include "WebExtensionContentWorldType.h"
+#include "WebExtensionTabIdentifier.h"
```
You might be able to just forward-declare WebExtensionTabIdentifier, which would reduce header fan-out when parsing.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    if (!hasPermission(webView.URL)) {
+        callback();
+        return;
+    }
```
```
auto tab = getTab(parameters.identifier.value());
auto *webView = tab->mainWebView();
// TODO: Should this return an error is tab isn't found?
// TODO: webView can also be nil. So return early after calling completionHandler.
if (!hasPermission(webView.URL, tab.get())) // need to pass tab for activeTab support.
    completionHandler();
    return;
}
```

xeenon replied:
Should also `ASSERT(parameters.identifier)`.

kiaraarose replied:
> // TODO: Should this return an error is tab isn't found?

In the WebProcess, I check if  the identifier returned from `toWebExtensionTabIdentifier` is valid and set *outExceptionString if it isn't. is that sufficient or would I need to do an additional check here?

kiaraarose replied:
Ah, but I guess `toWebExtensionTabIdentifier` only validates the format and not if it's actual tab. Will return an error here.

brentfulgham commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
```diff
@@ -28,6 +28,8 @@
 
 #import <JavaScriptCore/JSBase.h>
+#import <wtf/text/WTFString.h>
+#import <wtf/Vector.h>
```
Ideally, we would avoid including Vector and WTFString in this header. If you aren't declaring members of type Vector or String, it would be better to use `#include <wtf/Forward.h>`, and only include `Vector.h` and `WTFString.h` in the files that need them.

rreno commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+    return SourcePair([[NSString alloc] initWithData:scriptData encoding:NSUTF8StringEncoding], resourceURL);
+}
+
+void populateStyleSheetPairs(StyleSheetPairs&& pairs, RefPtr<WebExtension> extension, std::optional<Vector<String>> files, std::optional<String> css)
```
Ah I see whats going on. You want to take `pairs` by mutable reference, not by rvalue ref.
Even better: return the StyleSheetPairs from this function instead of taking as a parameter and mutating.

rreno replied:
Also ditto the std::optional<String> comment from before.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    if (!hasPermission(webView.URL)) {
+        callback(std::nullopt);
+        return;
+    }
```
See comment for `scriptingRemoveCSS`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+            WKFrameInfo *info = frame.get().info;
+            NSURL *frameURL = info.request.URL;
+
+            if (!hasPermission(frameURL)) {
```
`if (!hasPermission(frameURL, tab.get())`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+void WebExtensionContext::scriptingInsertCSS(WebPageProxyIdentifier identifier, const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void()>&& callback)
+{
+    WKWebView *webView = m_tabMap.get(parameters.identifier.value())->mainWebView();
+    if (!hasPermission(webView.URL)) {
```
See comment for `scriptingRemoveCSS`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    }).get()];
+}
+
+void WebExtensionContext::scriptingInsertCSS(WebPageProxyIdentifier identifier, const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void()>&& callback)
```
`CompletionHandler<void()>&& completionHandler`

Also should these take a `std::options<String> error` for when the tab isn't found, etc?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    callback();
+}
+
+void WebExtensionContext::scriptingRemoveCSS(WebPageProxyIdentifier identifier, const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void()>&& callback)
```
`CompletionHandler<void()>&& completionHandler`

Also should these take a `std::options<String> error` for when the tab isn't found, etc?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+using SourcePair = DynamicScripts::SourcePair;
+using ScriptVector = DynamicScripts::StyleSheetPairs;
+
+void WebExtensionContext::scriptingExecuteScript(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
```
`completionHandler` instead of `callback`.

Also should this take a `std::options<String> error` for when the tab isn't found, etc?

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+        size_t numberOfExecutions = frames.size();
+        Vector<WebExtensionScriptInjectionResultParameters> results;
+
+        for (auto& frame : frames) {
```
This seems wrong. Didn't you `WTFMove` the frames variable into `populateFrames`? That should mean that it is now empty. Whatever happens to `frames` inside of `populateFrames` will be invisible up at this layer of the stack. I think `frames` should just be passed as a reference so that we can observe its state after the call to `populateFrames`.

It actually might be a bit clearer if `populateFrames` just returned the populated Vector, since you don't seem to ever want to pass frames from elsewhere into that method.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+            executionWorld = API::ContentWorld::pageContentWorld();
+
+        Vector<RetainPtr<_WKFrameTreeNode>> frames;
+        if (parameters.allFrames || !parameters.frameIDs)
```
You might just be able to drop `allFrames` from parameters and use `!parameters.frameIDs` to mean all frames, that avoids the issues where they could be out off sync with each other.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+
+            if (!hasPermission(frameURL)) {
+                addResultAndCompleteIfDone(nil, info, toErrorString(@"scripting.executeScript()", nil, @"Extension does not have access to URL: %@", frameURL.absoluteString), numberOfExecutions, WTFMove(results), WTFMove(callback));
+                continue;
```
Once you hit `continue` here, you will loop around again. But you've just used `WTFMove` on the `results` and `callback` variables to pass their content to `addResultAndCompleteIfDone`, so now these symbols reference nothing. I think this might crash if you ever loop more than one time.

rreno commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -53,17 +57,91 @@
 static NSString * const mainWorld = @"MAIN";
 static NSString * const isolatedWorld = @"ISOLATED";
 
 static int64_t const mainFrameID = 0;
```
your static size_t mainFrameID defined in WebExtensionDynamicScripts.h might cause problems with thIs.

xeenon replied:
Yeah this should be removed too.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+
+namespace WebKit {
+
+namespace DynamicScripts {
```
`namespace  WebExtensionDynamicScripts` would likely be best, to avoid confusion. Not much issues since you use `using namespace DynamicScripts` already.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+                    scriptPairs.append(sourcePairForResource(file, m_extension));
+                
+                for (auto& script : scriptPairs) {
+                    [webView _evaluateJavaScript:script.value().first withSourceURL:script.value().second.value() inFrame:info inContentWorld:executionWorld.wrapper() completionHandler:makeBlockPtr([info = RetainPtr { info }, numberOfExecutions, results = WTFMove(results), callback = WTFMove(callback)](id resultOfJSExecution, NSError *error) mutable {
```
This is also suspicious. You call `_evaluateJavaScript` for each element in `scriptPairs`, but you are using `WTFMove` on the callback each loop. That doesn't seem right.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebKit {
+
+namespace DynamicScripts {
```
Newline below.

rreno commented on Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
```diff
@@ -0,0 +1,60 @@
+using SourcePair = std::pair<String, std::optional<URL>>;
+using StyleSheetPairs = Vector<std::optional<SourcePair>>;
+
+static const size_t mainFrameID = 0;
```
I'd be very careful defining constants in a header. Is there some other place we can get this magic value. Also is this even the correct ID? I think 0 is a placeholder ID rather than the main frame ID.

xeenon replied:
This is in a new namespace, but this is already defined as `WebExtensionFrameConstants::MainFrame` so this can be removed.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+            NSArray *arguments = parameters.args ? createNSArray(parameters.args.value()).get() : @[ ];
+            NSString *function = [NSString stringWithFormat:@"return (%@)(...args)", (NSString *)(parameters.func.value())];
+
+            [webView _callAsyncJavaScript:function arguments:@{ @"args": arguments } inFrame:info inContentWorld:executionWorld.wrapper() completionHandler:makeBlockPtr([info = RetainPtr { info }, numberOfExecutions, results = WTFMove(results), callback = WTFMove(callback)](id resultOfJSExecution, NSError *error) mutable {
```
Ditto: More moving of the `callback` and `results` variables for each step in the loop. Maybe `Arc` does something magic to keep these things alive somehow, but its confusing in the context of basic C++ operations like `WTFMove`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+namespace WebKit {
+
+struct WebExtensionScriptInjectionParameters {
+    std::optional<WebKit::WebExtensionTabIdentifier> identifier;
```
Would say `tabIdentifier` for clarity.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -86,6 +87,12 @@ T *objectForKey(const RetainPtr<NSDictionary>& dictionary, id key, bool returnin
     return objectForKey<T>(dictionary.get(), key, returningNilIfEmpty, containingObjectsOfClass);
 }
 
+inline bool boolForKey(NSDictionary *dictionary, id key, std::optional<bool> defaultValue)
```
Can be just `bool defaultValue`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+struct WebExtensionScriptInjectionParameters {
+    std::optional<WebKit::WebExtensionTabIdentifier> identifier;
+
+    std::optional<Vector<String>> args;
```
I'd spell this out. No need to propagate the bad API design internally.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+    std::optional<Vector<size_t>> frameIDs;
+
+    std::optional<String> css;
+    std::optional<String> func;
```
Ditto `function`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    auto page = WebProcessProxy::webPage(identifier);
+    WebUserContentControllerProxy& controller = page.get()->userContentController();
```
Can use auto more `auto& userContentController`

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -275,6 +275,14 @@
             continue;
 
+        for (auto& styleSheet : context->dynamicallyInjectedUserStyleSheets()) {
+            auto page = WebProcessProxy::webPage(pageID);
```
The return value of `WebProcessProxy:webPage()` might be a nullptr. You should either RELEASE_ASSERT that this will never happen, or do a nullptr check here (which is the typical pattern for this return type).

brentfulgham replied:
Actually, `pageID` doesn't change in this loop, so it would be better to get its value outside the loop, and just use the resulting Controller inside.

E.g.,
```
if (auto* page = WebProcessProxy::webPage(pageID)) {
    WebUserContentControllerProxy& controller = page.get()->userContentController();
    for (auto& styleSheet : context->dynamicallyInjectedUserStyleSheets())
        controller.removeUserStyleSheet(styleSheet);
}
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIScripting.idl:
```diff
@@ -30,8 +30,8 @@
-    [RaisesException] void insertCSS([NSDictionary] any details, [Optional, CallbackHandler] function callback);
-    [RaisesException] void removeCSS([NSDictionary] any details, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsPage] void insertCSS([NSDictionary] any details, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsPage] void removeCSS([NSDictionary] any details, [Optional, CallbackHandler] function callback);
```
Weird that `executeScript` does not need this. Why do these allow implicit tab and executeScript does not? Might be worth mention in the WECG.

kiaraarose replied:
In hindsight, it doesn't need it and this is wrong. I passed in the page here so that I could get the user content controller, but I should be getting the controllers from allUserContentControllers() from the extensions controller

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
```diff
@@ -54,6 +54,7 @@ class WebExtensionAPIScripting : public WebExtensionAPIObject, public JSWebExten
     bool validateScript(NSDictionary *, NSString **outExceptionString);
     bool validateTarget(NSDictionary *, NSString **outExceptionString);
     bool validateCSS(NSDictionary *, NSString **outExceptionString);
+
```
Remove.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+        foundStartingFrame = true;
+    }
+
+    for (_WKFrameTreeNode *child in currentNode.childFrames)
```
You possibly moved `currentNode` into `frames` on line 69, but then try to access its `childFrames` member here. Maybe the compiler is generating a RetainPtr on line 69, and moving that retain pointer, but it's confusing to read and I suggest you make it more explicit.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
+    if (frameIDs && frameIDs->contains(currentFrameID))
+        frames.append(WTFMove(currentNode));
```
I don't think you should have the `WTFMove` here. At best it's redundant, at worst it actually causes the `currentNode` pointer to be cleared.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
```diff
@@ -0,0 +1,60 @@
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#include "WebExtension.h"
+#include "WebExtensionScriptInjectionResultParameters.h"
```
I think you could just forward-declare `WebExtensionScriptInjectionResultParameters` here, and include in the implementation file.

brentfulgham replied:
Ditto for `WebExtensions`. You only use them as arguments to the methods, the compiler doesn't need to know the full type here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+        callback(results);
+}
+
+void populateFrames(_WKFrameTreeNode *currentNode, size_t startingFrameID, bool allFrames, bool foundStartingFrame, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
```
Need to use `uint64_t` for all frameID vars. `uint64_t startingFrameID`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+        frames.append(WTFMove(currentNode));
+
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
```
`uint64_t`

xeenon replied:
Ditto on `toWebExtensionFrameIdentifier`.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -53,17 +57,91 @@
 
 namespace WebKit {
 
+static inline NSArray *toAPI(Vector<WebExtensionScriptInjectionResultParameters> results)
```
Please pass `results` as a `const` reference to avoid copying the Vector.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+void populateFrames(_WKFrameTreeNode *currentNode, std::optional<Vector<size_t>> frameIDs, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
+{
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
```
`uint64_t`

xeenon replied:
Ditto on `toWebExtensionFrameIdentifier`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
```diff
@@ -0,0 +1,52 @@
+
+    std::optional<Vector<String>> args;
+    std::optional<Vector<String>> files;
+    std::optional<Vector<size_t>> frameIDs;
```
`uint64_t`

xeenon replied:
This should be `std::optional<Vector<WebExtensionFrameIdentifier>> frameIDs;`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionResultParameters.h:
```diff
@@ -0,0 +1,42 @@
+struct WebExtensionScriptInjectionResultParameters {
+    std::optional<String> error;
+    std::optional<String> result;
+    std::optional<size_t> frameID;
```
`uint64_t`

xeenon replied:
Actually `WebExtensionFrameIdentifier` also.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -78,31 +156,50 @@
 
-void WebExtensionAPIScripting::insertCSS(NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+void WebExtensionAPIScripting::insertCSS(WebPage* page, NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
 {
```
You pass `WebPage` in as a pointer, but never null-check it. If it's required that it is always non-null, please pass a reference. If it might be nullptr, please check for that.

brentfulgham replied:
Alternatively, consider just passing the WebPageProxy identifier, which seem to be the only thing you need from it.

xeenon replied:
This is called by the code generated code. It is pretty guaranteed to not be null since this comes from the JS world owned by the page, but I'll fix the code generator to pass references here instead. We do need the page sometimes, so it can't just be identifier.

kiaraarose replied:
We actually don't need the WebPage in this case, so I'll remove.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -78,31 +156,50 @@
 }
 
-void WebExtensionAPIScripting::removeCSS(NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+void WebExtensionAPIScripting::removeCSS(WebPage* page, NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
Ditto lack of `page` null check, or perhaps consider just passing the WebPageProxy identifier, which seem to be the only thing you need from it.

brentfulgham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -87,16 +87,17 @@
+//
+//        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, allFrames: false }, files: ['path/to/file'] }))",
+//        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, frameIds: [0] }, css: 'body { background-color: pink }' }))",
+//        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, allFrames: true }, css: 'body { background-color: pink }' }))",
```
Nit: We don't check commented-out code into WebKit. These tests should either be removed until you are ready, or these should be uncommented. (Or, mark them as expected to fail!)

kiaraarose replied:
Was planning on updating the PR with those tests as mentioned in the FIXME. Just wanted to get the PR up for review.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -53,17 +57,91 @@
+    if (NSArray *frameIds = targetInfo[frameIDsKey]) {
+        Vector<size_t> frames;
+        for (NSNumber *frameId in frameIds)
+            frames.append(frameId.unsignedLongLongValue);
```
This should really use `toWebExtensionFrameIdentifier()`, and those an exception if any come back as invalid.

brentfulgham reviewed (changes requested):
I think this generally looks good, but I think there are some mistakes in your use of `move` operations.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+{
+    WebExtensionScriptInjectionResultParameters result;
+    result.result = resultOfJSExecution ?: @"";
+    result.frameID = frame.isMainFrame ? mainFrameID : frame._handle.frameID;
```
I would make a helper in WebExtensionFrameIdentifier.h that `toWebExtensionFrameIdentifier(WKFrameInfo *)` and returns a `WebExtensionFrameIdentifier`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,188 @@
+    populateStyleSheetPairs(WTFMove(styleSheetPairs), m_extension, parameters.files, parameters.css);
+
+    for (auto& styleSheet : styleSheetPairs) {
+        auto userStyleSheet = API::UserStyleSheet::create(WebCore::UserStyleSheet { styleSheet.value().first, styleSheet.value().second.value_or(URL(""_s)), Vector<String>(), Vector<String>(), injectedFrames, WebCore::UserStyleUserLevel, std::nullopt }, *m_contentScriptWorld);
```
`URL { }` or just `{ }` should work.

xeenon replied:
`Vector<String> { }` is preferred over `Vector<String>()`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    if (!hasPermission(webView.URL, tab.get())) {
+        completionHandler(std::nullopt, toErrorString(callingAPIName, nil, @"Extension does not have access to URL: %@", webView.URL.absoluteString));
```
This error is reported back to the extension, and will leak the URL to them. We should just say "this extension does not have access in this tab".

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    if (!hasPermission(webView.URL, tab.get())) {
+        completionHandler(toErrorString(callingAPIName, nil, @"Extension does not have access to URL: %@", webView.URL.absoluteString));
```
Ditto.  "this extension does not have access in this tab".

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    if (!hasPermission(webView.URL, tab.get())) {
+        completionHandler(toErrorString(callingAPIName, nil, @"Extension does not have access to URL: %@", webView.URL.absoluteString));
```
Ditto. "this extension does not have access in this tab".

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+        return;
+    }
+
+    NSString *callingAPIName = @"scripting.removeCSS()";
```
`static NSString * const apiName = ...`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+
+void WebExtensionContext::scriptingInsertCSS(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(WebExtensionDynamicScripts::Error)>&& completionHandler)
+{
+    NSString *callingAPIName = @"scripting.insertCSS()";
```
`static NSString * const apiName = ...`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+
+void WebExtensionContext::scriptingExecuteScript(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<InjectionResults>, WebExtensionDynamicScripts::Error)>&& completionHandler)
+{
+    NSString *callingAPIName = @"scripting.executeScript()";
```
`static NSString * const apiName = ...`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+void WebExtensionContext::scriptingInsertCSS(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(WebExtensionDynamicScripts::Error)>&& completionHandler)
+{
+    NSString *callingAPIName = @"scripting.insertCSS()";
+    auto tab = getTab(parameters.tabIdentifier.value());
```
Needs `tab not found` error check.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    NSString *callingAPIName = @"scripting.removeCSS()";
+    auto tab = getTab(parameters.tabIdentifier.value());
```
Needs `tab not found` error check.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    auto injectedFrames = parameters.frameIDs ? WebCore::UserContentInjectedFrames::InjectInTopFrameOnly : WebCore::UserContentInjectedFrames::InjectInAllFrames;
+
+    UserStyleSheetVector styleSheetsToRemove;
+    Vector<std::optional<SourcePair>> styleSheetPairs = populateStyleSheetPairs(m_extension, parameters.files, parameters.css);
```
Can be `auto styleSheetPairs = …`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    auto injectedFrames = parameters.frameIDs ? WebCore::UserContentInjectedFrames::InjectInTopFrameOnly : WebCore::UserContentInjectedFrames::InjectInAllFrames;
+
+    SourcePairs styleSheetPairs = populateStyleSheetPairs(m_extension, parameters.files, parameters.css);
+    if (auto pageID = webView._page->webPageID()) {
```
No need to null check it with the if.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,166 @@
+    Vector<std::optional<SourcePair>> styleSheetPairs = populateStyleSheetPairs(m_extension, parameters.files, parameters.css);
+
+    for (auto& styleSheetContent : styleSheetPairs) {
+        for (auto userStyleSheet : m_dynamicallyInjectedUserStyleSheets) {
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,101 @@
+    if (!scriptData)
+        return std::nullopt;
+
+    URL resourceURL = URL(extension->resourceFileURLForPath(path));
```
`auto resourceURL = URL { extension->resourceFileURLForPath(path) };`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,101 @@
+
+std::optional<SourcePair> sourcePairForResource(String path, RefPtr<WebExtension> extension)
+{
+    NSData *scriptData = extension->resourceDataForPath(path);
```
`auto *scriptData = ...`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,101 @@
+        return styleSheets;
+    }
+
+    return { make_pair(css.value(), std::nullopt) };
```
`return { SourcePair { css.value(), std::nullopt } };`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,101 @@
+        return std::nullopt;
+
+    URL resourceURL = URL(extension->resourceFileURLForPath(path));
+    return SourcePair([[NSString alloc] initWithData:scriptData encoding:NSUTF8StringEncoding], resourceURL);
```
`return SourcePair { [[NSString alloc] initWithData:scriptData encoding:NSUTF8StringEncoding], resourceURL };`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -0,0 +1,101 @@
+    return matchingFrames;
+}
+
+std::optional<SourcePair> sourcePairForResource(String path, RefPtr<WebExtension> extension)
```
I'd put the extension arg first, to match populateStyleSheetPairs, or vice versa.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -53,17 +59,76 @@
+    }
+
+    if (!boolForKey(targetInfo, allFramesKey, false))
+        parameters.frameIDs = Vector<WebExtensionFrameIdentifier> { WebExtensionFrameConstants::MainFrameIdentifier };
```
I think `... = Vector { WebExtensionFrameConstants::MainFrameIdentifier };` will work, since the template can be inferred by the type in the `{ }`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,146 @@
+
+    auto *webView = tab->mainWebView();
+    if (!webView) {
+        completionHandler(std::nullopt, toErrorString(apiName, nil, @"could not execute script to this tab"));
```
"could not execute script *on* this tab"

kiaraarose committed:
Work towards implementing browsing.scripting() web extension APIs.
https://bugs.webkit.org/show_bug.cgi?id=259954

Reviewed by Timothy Hatcher.

This patch adds support for script injections with the scripting.insertCSS(), and scripting.removeCSS() APIs.

* Source/WTF/wtf/Ref.h:
(WTF::operator==):
Add support for operator== so that removeFirst(const U&) works for Ref types.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
(WebKit::boolForKey):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h:
(WebKit::toWebExtensionFrameIdentifier):
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionResultParameters.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm: Added.
(WebKit::WebExtensionContext::scriptingExecuteScript):
(WebKit::WebExtensionContext::scriptingInsertCSS):
(WebKit::WebExtensionContext::scriptingRemoveCSS):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didCommitLoadForFrame):
Clear dynamicallyInjectedUserStyleSheets when a page navigation occurs since the stylesheets should
be removed when the user leaves or refreshes the page.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm: Added.
(WebKit::WebExtensionDynamicScripts::userStyleSheetMatchesContent):
(WebKit::WebExtensionDynamicScripts::populateFrames):
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
(WebKit::WebExtensionDynamicScripts::populateStyleSheetPairs):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::dynamicallyInjectedUserStyleSheets):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::parseTargetInjectionOptions):
(WebKit::parseScriptInjectionOptions):
(WebKit::parseCSSInjectionOptions):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::validateTarget):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/269018@main
